File: agent/dotnet/src/Perper.Protocol/StreamData.cs
using System;
using System.Collections;
using System.Linq;
using System.Collections.Generic;
using System.Threading.Tasks;
using Apache.Ignite.Core;
using Apache.Ignite.Core.Binary;
using Apache.Ignite.Core.Cache.Affinity;
using Apache.Ignite.Core.Client;

namespace Perper.Protocol
{
    public static class StreamData
    {

        public static IBinaryObject CreateStreamData<TParams>(
            IBinary binary,
            string instance,
            string agent,
            string @delegate,
            StreamDelegateType delegateType,
            bool ephemeral,
            TParams parameters,
            string? indexType = null,
            Hashtable? indexFields = null)
        {
            var builder = binary.GetBuilder($"StreamData_{agent}_{@delegate}");

            builder.SetField("instance", instance);
            builder.SetField("agent", agent);
            builder.SetField("delegate", @delegate);
            builder.SetField("delegateType", delegateType);
            builder.SetField("ephemeral", ephemeral);
            builder.SetField("listeners", new ArrayList());
            builder.SetField("indexFields", indexFields);
            builder.SetField("indexType", indexType);
            builder.SetField("parameters", parameters);

            return builder.Build();
        }

        /*public class StreamListener
        {
            public string callerAgent { get; set; }
            public string caller { get; set; }
            public int parameter { get; set; }
            public Hashtable filter { get; set; }
            public bool replay { get; set; }
            public bool localToData { get; set; }
        }*/

        public static IBinaryObject CreateStreamListener(
            IBinary binary,
            string callerAgent,
            string caller,
            int parameter,
            bool replay,
            bool localToData,
            Hashtable? filter = null)
        {
            /*return binary.ToBinary<IBinaryObject>(new StreamListener
            {
                callerAgent = callerAgent,
                caller = caller,
                parameter = parameter,
                filter = filter,
                replay = replay,
                localToData = localToData
            });*/

            var builder = binary.GetBuilder("StreamListener");

            builder.SetField("callerAgent", callerAgent);
            builder.SetField("caller", caller);
            builder.SetField("parameter", parameter);
            builder.SetField("filter", filter ?? new Hashtable());
            builder.SetField("replay", replay);
            builder.SetField("localToData", localToData);

            return builder.Build();
        }

        public static IBinaryObject StreamDataAddListener(
            IBinaryObject streamData,
            IBinaryObject listener)
        {
            var builder = streamData.ToBuilder();

            var listeners = builder.GetField<ArrayList>("listeners");
            listeners.Add(listener.ToBuilder()); // FIXME: workaround for a weird schema-registration issue in Ignite
            builder.SetField("listeners", listeners);

            return builder.Build();
        }

        public static IBinaryObject StreamDataRemoveListener(
            IBinaryObject streamData,
            IBinaryObject listener)
        {
            var builder = streamData.ToBuilder();

            var listeners = builder.GetField<ArrayList>("listeners");
            listeners.Remove(listener);
            builder.SetField("listeners", listeners);

            return builder.Build();
        }
    }
}

File: agent/python/perper/stream_data.py
from collections import OrderedDict
from pyignite import GenericObjectMeta
from pyignite.utils import entity_id
from pyignite.datatypes import String, IntObject, BoolObject, MapObject, EnumObject, CollectionObject

StreamDelegateTypeId = entity_id("StreamDelegateType")

def create_stream_data(instance, agent, delegate, delegateType, ephemeral, parameters, parametersType, indexType=None, indexFields=None):
    class StreamData(metaclass=GenericObjectMeta, type_name=f"StreamData_{agent}_{delegate}", schema=OrderedDict([
        ('instance', String),
        ('agent', String),
        ('delegate', String),
        ('delegateType', EnumObject),
        ('indexType', String),
        ('indexFields', MapObject),
        ('ephemeral', BoolObject),
        ('listeners', CollectionObject),
        ('parameters', parametersType),
    ])):
        pass

    return StreamData(
        instance=instance,
        agent=agent,
        delegate=delegate,
        delegateType=(StreamDelegateTypeId, delegateType),
        indexType=indexType,
        indexFields=None if indexFields is None else (MapObject.HASH_MAP, indexFields),
        ephemeral=ephemeral,
        listeners=(CollectionObject.ARR_LIST, []),
        parameters=parameters
    )

class StreamListener(metaclass=GenericObjectMeta, schema=OrderedDict([
    ('callerAgent', String),
    ('caller', String),
    ('parameter', IntObject),
    ('filter', MapObject),
    ('replay', BoolObject),
    ('localToData', BoolObject),
])):
    pass

def create_stream_listener(callerAgent, caller, parameter, replay, localToData, filter={}):
    return StreamListener(
        callerAgent=callerAgent,
        caller=caller,
        parameter=parameter,
        replay=replay,
        localToData=localToData,
        filter=(MapObject.HASH_MAP, filter)
    )

def stream_data_add_listener(streamData, streamListener):
    StreamData = type(streamData)

    new_listeners = streamData.listeners[1][:]
    new_listeners.append(streamListener)

    return StreamData(
        listeners=(CollectionObject.ARR_LIST, new_listeners),
        **{key: getattr(streamData, key) for key in streamData._schema.keys() - {'listeners'}}
    )

def stream_data_remove_listener(streamData, streamListener):
    StreamData = type(streamData)

    new_listeners = streamData.listeners[1][:]
    new_listeners.remove(streamListener)

    return StreamData(
        listeners=(CollectionObject.ARR_LIST, new_listeners),
        **{key: getattr(streamData, key) for key in streamData._schema.keys() - {'listeners'}}
    )
